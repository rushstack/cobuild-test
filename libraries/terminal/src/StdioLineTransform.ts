// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Text, NewlineKind } from '@rushstack/node-core-library';

import { ITerminalChunk, TerminalChunkKind } from './ITerminalChunk';
import { TerminalTransform, ITerminalTransformOptions } from './TerminalTransform';

/**
 * Constructor options for {@link StderrLineTransform}
 * @beta
 */
export interface IStdioLineTransformOptions extends ITerminalTransformOptions {
  /**
   * Specifies the kind of newline for the output.
   */
  newlineKind?: NewlineKind;
}

/**
 * `StderrLineTransform` normalizes lines that mix characters from `stdout` and `stderr`,
 * so that each output line is routed entirely to `stdout` or `stderr`.
 *
 * @remarks
 * IMPORTANT: This transform assumes that its input has been normalized to use `"\n"` newlines.
 *
 * IMPORTANT: This transform does not produce realtime output, because lines are buffered
 * until a newline character is encountered.
 *
 * Suppose that a poorly behaved process produces output like this:
 *
 * ```ts
 * process.stderr.write('An error occurred, cleaning up');
 * process.stdout.write('.'); // (delay)
 * process.stdout.write('.'); // (delay)
 * process.stdout.write('.');
 * process.stdout.write('\n');
 * process.stderr.write('The process completed with errors\n');
 * ```
 *
 * When `stdout` and `stderr` are combined on the console, the mistake in the output would not be noticeable:
 * ```
 * An error occurred, cleaning up...
 * The process completed with errors
 * ```
 *
 * However, if we discard `stdout`, then `stderr` is malformed:
 * ```
 * An error occurred, cleaning upThe process completed with errors
 * ```
 *
 * Tooling scripts can introduce these sorts of problems via edge cases that are difficult to find and fix.
 *
 * `StderrLineTransform` normalizes the output so that if a combined line contains any `stderr` characters,
 * then the entire line is routed to `stderr`.  Later, if we discard `stdout`, then the output will
 * preserve the appropriate context:
 *
 * ```
 * An error occurred, cleaning up...
 * The process completed with errors
 * ```
 *
 * @privateRemarks
 * This class is experimental and marked as `@beta`.  The algorithm may need some fine-tuning, or there may
 * be better solutions to this problem.
 *
 * @beta
 */
export class StderrLineTransform extends TerminalTransform {
  private _accumulatedLine: string;
  private _accumulatedStderr: boolean;

  public readonly newline: string;

  public constructor(options: IStdioLineTransformOptions) {
    super(options);

    this._accumulatedLine = '';
    this._accumulatedStderr = false;

    this.newline = Text.getNewline(options.newlineKind || NewlineKind.Lf);
  }

  protected onWriteChunk(chunk: ITerminalChunk): void {
    if (chunk.text.indexOf('\r') >= 0) {
      throw new Error('StderrLineTransform expects chunks with normalized newlines');
    }

    // After _newlineNormalizerTransform was applied, we can now assume that all newlines
    // use the "\n" string
    const text: string = chunk.text;
    let startIndex: number = 0;

    while (startIndex < text.length) {
      if (chunk.kind === TerminalChunkKind.Stderr) {
        this._accumulatedStderr = true;
      }

      const endIndex: number = text.indexOf('\n', startIndex);
      if (endIndex < 0) {
        // we did not find \n, so simply append
        this._accumulatedLine += text.substring(startIndex);
        break;
      }

      // append everything up to \n
      this._accumulatedLine += text.substring(startIndex, endIndex);

      this._processAccumulatedLine();

      // skip the \n
      startIndex = endIndex + 1;
    }
  }

  protected onClose(): void {
    if (this._accumulatedLine.length > 0) {
      this._processAccumulatedLine();
    }
    this.autocloseDestination();
  }

  private _processAccumulatedLine(): void {
    this._accumulatedLine += this.newline;

    if (this._accumulatedStderr) {
      this.destination.writeChunk({
        kind: TerminalChunkKind.Stderr,
        text: this._accumulatedLine
      });
    } else {
      this.destination.writeChunk({
        kind: TerminalChunkKind.Stdout,
        text: this._accumulatedLine
      });
    }

    this._accumulatedLine = '';
    this._accumulatedStderr = false;
  }
}
