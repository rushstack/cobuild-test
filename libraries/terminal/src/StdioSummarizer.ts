// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { ITerminalChunk, TerminalChunkKind } from './ITerminalChunk';
import { TerminalWritable } from './TerminalWritable';

/**
 * Constructor options for {@link StdioSummarizer}.
 * @beta
 */
export interface IStdioSummarizerOptions {
  /**
   * Specifies the maximum number of leading lines to include in the summary.
   * @defaultValue `10`
   */
  leadingLines?: number;

  /**
   * Specifies the maximum number of trailing lines to include in the summary.
   * @defaultValue `10`
   */
  trailingLines?: number;
}

/**
 * Summarizes the results of a failed build task by returning a subset of `stderr` output not to exceed
 * a specified maximum number of lines.
 *
 * @remarks
 * IMPORTANT: This transform assumes that its input was prepared by {@link StderrLineTransform}, so that each
 * {@link ITerminalChunk.text} item is a single line terminated by a `"\n"` character.
 *
 * The {@link IStdioSummarizerOptions.leadingLines} and {@link IStdioSummarizerOptions.trailingLines}
 * counts specify the maximum number of lines to be returned. Any additional lines will be omitted.
 * For example, if `leadingLines` and `trailingLines` were set to `3`, then the summary of 16 `stderr` lines might
 * look like this:
 *
 * ```
 * Line 1
 * Line 2
 * Line 3
 *   ...10 lines omitted...
 * Line 14
 * Line 15
 * Line 16
 * ```
 *
 * If the `stderr` output is completely empty, then the `stdout` output will be summarized instead.
 *
 * @beta
 */
export class StdioSummarizer extends TerminalWritable {
  // Capture up to this many leading lines
  private _leadingLines: number;

  // Capture up to this many trailing lines
  private _trailingLines: number;

  private readonly _abridgedLeading: string[];
  private readonly _abridgedTrailing: string[];
  private _abridgedOmittedLines: number = 0;
  private _abridgedStderr: boolean;

  public constructor(options?: IStdioSummarizerOptions) {
    super();

    if (!options) {
      options = {};
    }

    this._leadingLines = options.leadingLines !== undefined ? options.leadingLines : 10;
    this._trailingLines = options.trailingLines !== undefined ? options.trailingLines : 10;

    this._abridgedLeading = [];
    this._abridgedTrailing = [];
    this._abridgedStderr = false;
  }

  /**
   * Returns the summary report.
   *
   * @remarks
   * The `close()` method must be called before `getReport()` can be used.
   */
  public getReport(): string {
    if (this.isOpen) {
      throw new Error('The summary cannot be prepared until after close() is called.');
    }
    const report: string[] = [...this._abridgedLeading];
    if (this._abridgedOmittedLines === 1) {
      report.push(`  ...${this._abridgedOmittedLines} line omitted...\n`);
    }
    if (this._abridgedOmittedLines > 1) {
      report.push(`  ...${this._abridgedOmittedLines} lines omitted...\n`);
    }
    report.push(...this._abridgedTrailing);
    return report.join('');
  }

  public onWriteChunk(chunk: ITerminalChunk): void {
    if (chunk.text.length === 0 || chunk.text[chunk.text.length - 1] !== '\n') {
      throw new Error(
        'StdioSummarizer expects chunks that were separated parsed into lines by StderrLineTransform\n' +
          ' Invalid input: ' +
          JSON.stringify(chunk.text)
      );
    }

    if (chunk.kind === TerminalChunkKind.Stderr && !this._abridgedStderr) {
      // The first time we see stderr, switch to capturing stderr
      this._abridgedStderr = true;
      this._abridgedLeading.length = 0;
      this._abridgedTrailing.length = 0;
      this._abridgedOmittedLines = 0;
    } else if (this._abridgedStderr && chunk.kind !== TerminalChunkKind.Stderr) {
      // If we're capturing stderr, then ignore non-stderr input
      return;
    }

    // Did we capture enough leading lines?
    if (this._abridgedLeading.length < this._leadingLines) {
      this._abridgedLeading.push(chunk.text);
      return;
    }

    this._abridgedTrailing.push(chunk.text);

    // If we captured to many trailing lines, omit the extras
    while (this._abridgedTrailing.length > this._trailingLines) {
      this._abridgedTrailing.shift();
      ++this._abridgedOmittedLines;
    }
  }
}
