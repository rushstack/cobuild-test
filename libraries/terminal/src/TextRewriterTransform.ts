// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { NewlineKind } from '@rushstack/node-core-library';

import { ITerminalChunk, TerminalChunkKind } from './ITerminalChunk';
import { TerminalTransform, ITerminalTransformOptions } from './TerminalTransform';
import { TextRewriter, TextRewriterState } from './TextRewriter';
import { RemoveColorsTextRewriter } from './RemoveColorsTextRewriter';
import { NormalizeNewlinesTextRewriter } from './NormalizeNewlinesTextRewriter';

/**
 * Constructor options for {@link TextRewriterTransform}.
 *
 * @public
 */
export interface ITextRewriterTransformOptions extends ITerminalTransformOptions {
  /**
   * A list of rewriters to be applied.  More items may be appended to the list, for example
   * if {@link ITextRewriterTransformOptions.removeColors} is specified.
   *
   * @remarks
   * The final list must contain at least one item.
   */
  textRewriters?: TextRewriter[];

  /**
   * If specified, a {@link RemoveColorsTextRewriter} will be appended to the list of rewriters.
   */
  removeColors?: boolean;

  /**
   * If `normalizeNewlines` or `ensureNewlineAtEnd` is specified, a {@link NormalizeNewlinesTextRewriter}
   * will be appended to the list of rewriters with the specified settings.
   *
   * @remarks
   * See {@link INormalizeNewlinesTextRewriterOptions} for details.
   */
  normalizeNewlines?: NewlineKind;

  /**
   * If `normalizeNewlines` or `ensureNewlineAtEnd` is specified, a {@link NormalizeNewlinesTextRewriter}
   * will be appended to the list of rewriters with the specified settings.
   *
   * @remarks
   * See {@link INormalizeNewlinesTextRewriterOptions} for details.
   */
  ensureNewlineAtEnd?: boolean;
}

/**
 * A {@link TerminalTransform} subclass that performs one or more {@link TextRewriter} operations.
 * The most common operations are {@link NormalizeNewlinesTextRewriter} and {@link RemoveColorsTextRewriter}.
 *
 * @remarks
 * The `TextRewriter` operations are applied separately to the `stderr` and `stdout` streams.
 * If multiple {@link ITextRewriterTransformOptions.textRewriters} are configured, they are applied
 * in the order that they appear in the array.
 *
 * @public
 */
export class TextRewriterTransform extends TerminalTransform {
  private readonly _stderrStates: TextRewriterState[];
  private readonly _stdoutStates: TextRewriterState[];

  public readonly textRewriters: ReadonlyArray<TextRewriter>;

  public constructor(options: ITextRewriterTransformOptions) {
    super(options);

    const textRewriters: TextRewriter[] = options.textRewriters || [];

    if (options.removeColors) {
      textRewriters.push(new RemoveColorsTextRewriter());
    }
    if (options.normalizeNewlines) {
      textRewriters.push(
        new NormalizeNewlinesTextRewriter({
          newlineKind: options.normalizeNewlines,
          ensureNewlineAtEnd: options.ensureNewlineAtEnd
        })
      );
    }

    if (textRewriters.length === 0) {
      throw new Error('TextRewriterTransform requires at least one matcher');
    }

    this.textRewriters = textRewriters;

    this._stderrStates = this.textRewriters.map((x) => x.initialize());
    this._stdoutStates = this.textRewriters.map((x) => x.initialize());
  }

  protected onWriteChunk(chunk: ITerminalChunk): void {
    if (chunk.kind === TerminalChunkKind.Stderr) {
      this._processText(chunk, this._stderrStates);
    } else if (chunk.kind === TerminalChunkKind.Stdout) {
      this._processText(chunk, this._stdoutStates);
    } else {
      this.destination.writeChunk(chunk);
    }
  }

  private _processText(chunk: ITerminalChunk, states: TextRewriterState[]): void {
    let text: string = chunk.text;
    for (let i: number = 0; i < states.length; ++i) {
      if (text.length > 0) {
        text = this.textRewriters[i].process(states[i], text);
      }
    }
    if (text.length > 0) {
      // If possible, avoid allocating a new chunk
      if (text === chunk.text) {
        this.destination.writeChunk(chunk);
      } else {
        this.destination.writeChunk({
          text: text,
          kind: chunk.kind
        });
      }
    }
  }

  private _closeRewriters(states: TextRewriterState[], chunkKind: TerminalChunkKind): void {
    let text: string = '';
    for (let i: number = 0; i < states.length; ++i) {
      if (text.length > 0) {
        text = this.textRewriters[i].process(states[i], text);
      }
      text += this.textRewriters[i].close(states[i]);
    }
    if (text.length > 0) {
      this.destination.writeChunk({
        text: text,
        kind: chunkKind
      });
    }
  }

  protected onClose(): void {
    this._closeRewriters(this._stderrStates, TerminalChunkKind.Stderr);
    this._closeRewriters(this._stdoutStates, TerminalChunkKind.Stdout);

    this.autocloseDestination();
  }
}
