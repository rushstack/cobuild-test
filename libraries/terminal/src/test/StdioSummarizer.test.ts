// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { TerminalChunkKind } from '../ITerminalChunk';
import { StdioSummarizer } from '../StdioSummarizer';
import { StderrLineTransform } from '../StdioLineTransform';
import { TextRewriterTransform } from '../TextRewriterTransform';
import { NewlineKind } from '@rushstack/node-core-library';

describe(StdioSummarizer.name, () => {
  let summarizer: StdioSummarizer;
  let stderrLineTransform: StderrLineTransform;
  let transform: TextRewriterTransform;

  beforeEach(() => {
    summarizer = new StdioSummarizer();
    stderrLineTransform = new StderrLineTransform({ destination: summarizer });
    transform = new TextRewriterTransform({
      destination: stderrLineTransform,
      normalizeNewlines: NewlineKind.Lf
    });
  });

  it('should report stdout if there is no stderr', () => {
    transform.writeChunk({ text: 'stdout 1\nstdout 2\n', kind: TerminalChunkKind.Stdout });
    transform.close();

    expect(summarizer.isOpen).toBe(false);
    expect(summarizer.getReport()).toMatchSnapshot();
  });

  it('should abridge extra lines', () => {
    transform.writeChunk({ text: 'discarded stdout\n', kind: TerminalChunkKind.Stdout });
    for (let i: number = 0; i < 10; ++i) {
      transform.writeChunk({ text: `leading ${i}\n`, kind: TerminalChunkKind.Stderr });
      transform.writeChunk({ text: 'discarded stdout\n', kind: TerminalChunkKind.Stdout });
    }

    transform.writeChunk({ text: `discarded middle 1\n`, kind: TerminalChunkKind.Stderr });
    transform.writeChunk({ text: `discarded middle 2\n`, kind: TerminalChunkKind.Stderr });

    for (let i: number = 0; i < 10; ++i) {
      transform.writeChunk({ text: `trailing ${i}\n`, kind: TerminalChunkKind.Stderr });
      transform.writeChunk({ text: 'discarded stdout\n', kind: TerminalChunkKind.Stdout });
    }

    transform.close();

    expect(summarizer.getReport()).toMatchSnapshot();
  });

  it('should concatenate partial lines', () => {
    transform.writeChunk({ text: 'abc', kind: TerminalChunkKind.Stderr });
    transform.writeChunk({ text: '', kind: TerminalChunkKind.Stderr });
    transform.writeChunk({ text: 'de\nf\n\ng', kind: TerminalChunkKind.Stderr });
    transform.writeChunk({ text: '\n', kind: TerminalChunkKind.Stderr });
    transform.writeChunk({ text: 'h', kind: TerminalChunkKind.Stderr });
    transform.close();

    expect(summarizer.getReport()).toMatchSnapshot();
  });
});
