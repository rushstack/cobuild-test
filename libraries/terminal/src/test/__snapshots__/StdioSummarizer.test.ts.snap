// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StdioSummarizer should abridge extra lines 1`] = `
"leading 0
leading 1
leading 2
leading 3
leading 4
leading 5
leading 6
leading 7
leading 8
leading 9
  ...2 lines omitted...
trailing 0
trailing 1
trailing 2
trailing 3
trailing 4
trailing 5
trailing 6
trailing 7
trailing 8
trailing 9
"
`;

exports[`StdioSummarizer should concatenate partial lines 1`] = `
"abcde
f

g
h
"
`;

exports[`StdioSummarizer should report stdout if there is no stderr 1`] = `
"stdout 1
stdout 2
"
`;
